propka/run.py

```diff
@@ -36,34 +36,68 @@ def main(optargs=None):
+def single(filename: str, optargs: list, stream=None, write_pka: bool = True):
+    """Run a single PROPKA calculation using ``filename`` as input.
 
-    Commandline options can be passed as a **list** in *optargs*.
+    Args:
+        filename (str): name of input file. If filestream is not passed via
+            ``stream``, should be path to the file to be read.
+        optargs (list): Optional, commandline options for propka.
+        stream : optional filestream handle. If ``None``, then ``filename``
+            will be used as path to input file for reading.
+        write_pka (bool): Controls if the pKa file should be writen to disk.
 
-    Example
-    -------
+    Example:
     Given an input file "protein.pdb", run the equivalent of ``propka3
     --mutation=N25R/N181D -v --pH=7.2 protein.pdb`` as::
 
-       propka.run.single("protein.pdb",
-                         optargs=["--mutation=N25R/N181D", "-v", "--pH=7.2"])
+        propka.run.single("protein.pdb",
+                          optargs=["--mutation=N25R/N181D", "-v", "--pH=7.2"])
 
+    By default, a pKa file will be written. However in some cases one may wish
+    to not output this file and just have access to the
+    MolecularContainer object. If so, then pass ``False`` to ``write_pka``::
 
-    .. todo::
-       Test :func:`single`, not sure if it is correctly processing ``pdbfile``.
+        mol = propka.run.single("protein.pdb", write_pka=False)
+
+    In some cases, one may also want to pass a file-like (e.g. StringIO) object
+    instead of a file path as a string. In these cases the file-like object
+    should be passed to the ``stream`` argument. Since file-like objects do not
+    usually have names, and the ``filename`` argument must be provided to
+    help propka determine the input file type, and assigns the file name for
+    the MolecularContainer object::
+
+        mol = propka.run.single('input.pdb', stream=string_io_file)
+
+    In this case, a PDB file-like object was passed as `string_io_file`. The
+    resultant pKa file will be written out as `input.pka`.
 
     """
+    # Deal with input optarg options
     optargs = optargs if optargs is not None else []
-    options = loadOptions(*optargs)
-    pdbfile = options.filenames.pop(0)
+    optargs += [filename]
+    options = loadOptions(optargs)
+
     parameters = read_parameter_file(options.parameters, Parameters())
-    if len(options.filenames) > 0:
-        _LOGGER.warning("Ignoring filenames: {0:s}".format(options.filenames))
+
+    # Only filename present should be the one passed via the arguments list
+    # Anything else will probably have been passed using optargs and the `-f`
+    # flag. Note: can be mulitple -f entries and all are appended before the
+    # input filename
+    if len(options.filenames) > 1:
+        ignored_list = options.filenames[:-1]
+        # overwrite the options.filenames entry
+        options.filenames = options.filenames[-1]
```

Comment:
Note: overwriting `options.filenames` here so that MoleculeContainer.options contains the correct filenames entry.